src/Applications/LDAS_App/preprocess_ldas_routines.F90

```diff
@@ -2152,32 +2153,53 @@ subroutine optimize_latlon(fname,arg)
           n2 = k*IMGLOB
           
           do i=1,60
-             rates(i) = (i-1)*0.1
+             rates(i) = -0.3 + i*0.01
           enddo
           
           maxf=rms_cs(rates)
           i=minloc(maxf,DIM=1)
           rate = rates(i)
           avg_land = ceiling(1.0*face_land(k)/face(k))
-          avg_land = avg_land - nint(rate*avg_land/face(k))
+          avg_land = avg_land - nint(rate*avg_land)
           
           tmpint = 0
           j = j+face(k) 
-          
+          forward = .true.
           do n = n1,n2
              tmpint=tmpint+landPosition(n)
              if((local+1) == j .and. n < n2) cycle
-             if((tmpint .ge. avg_land) .or. (n==n2)) then
+             if(n==n2) then
                 local = local + 1
                 local_land(local)=tmpint
                 JMS(local)=n-n0
                 tmpint=0
                 n0=n
+                cycle
              endif
+             if(tmpint .ge. avg_land) then
+                local = local + 1
+                if (n-n0 == 1) forward =.true.
+                if (forward) then
+                   local_land(local)=tmpint
+                   JMS(local)=n-n0
+                   tmpint=0
+                   n0=n
+                   forward = .false.
+                else
+                   local_land(local)=tmpint - landPosition(n)
+                   JMS(local)=n-1-n0
+                   tmpint=landPosition(n)
+                   n0=n-1
+                   forward = .true.
+                endif
+            endif
           enddo
           local = j
        enddo
-       
+      if( sum(JMS) /= JMGLOB) then
+          print*, sum(JMS), JMGLOB
+          stop ("wrong cs-domain distribution in the first place")
+       endif 
        ! adjust JMS.rc make make sure no process has 0 or 1
```

Comment:
What happens if a user (inadvertently) requests more processors than tiles, or if they are running only a single (cube-sphere) tile on one processor?  In this case, there would have to be at least one processor with 0 or 1 tiles.  
I don't think anyone has run a single cube-sphere tile yet, but users occasionally run a single  EASE grid tile.
Unless I misunderstand the comments (and I didn't manage to follow the algorithm implementation yet for lack of time), we might need to allow the case of a single cubesphere tile on a single processor and also provide a more helpful stop message if the user requests more processors than there are tiles.

Reply:
To run cubed-sphere grid, the minimum number of processors is 6. Here the comment  "adjust JMS.rc make make sure no process has 0 or 1" is about the number of grid cells, not tiles.

Reply:
Ok, thanks, but I still don't understand exactly what is and isn't possible, and what the stop/exit messages are when something doesn't work.  Let me ask the question differently:
If a user wants to run a single tile from a cube-sphere tile space, do they have to request 6 processors?  
If a user wants to run a single tile from a cube-sphere tile space, would the program stop because for a single-tile domain there can never be at least two tiles on each processor (regardless of how many processors are requested)?
Sorry for being slow to catch on.

Reply:
This is sanity check. For example, if it is c72, after distribution, the sum of grid cells is not 72, then it stops.  1) yes, users need to have 6, 12, 18, ..... processors. Actually this is not our requirement, it is MAPL's requirement.  2) Again this is about grid cells, not tiles. For example, if it is c72, the number of processor for a face should not exceed 36 ( we divided the grid along one dimension). If it is bigger than 36, then some processors would end up with 1 grid cell.

Reply:
It is fine to run with one tile and 6 processors

Reply:
Corrected: If it is c72, after distribution, the sum of grid cells is not 6*72, the it stops

Reply:
Thanks, @weiyuan-jiang.  I tried my luck at adding some help text about the domain decomposition in this latest commit:
[0d376a0](https://github.com/GEOS-ESM/GEOSldas/pull/418/commits/0d376a083a8de82c1cbbb5e641976ab09c1bea27?w=1)
The commit also contains minor cleanup (incl. added "intent" statements)